feat: Support bubbling up of events from evented children on dataclasses

tlambert03 opened the pull request:
picking up from #169   ... starting over since lots has changed


```python
from psygnal import evented
from dataclasses import dataclass, field

@evented
@dataclass
class Foo:
    x: int = 1

@evented(connect_child_events=True)
@dataclass
class Bar:
    foo: Foo = field(default_factory=Foo)


bar = Bar()
bar.events.connect(print)
bar.foo.x = 8
```

prints

```python
EmissionInfo(
    signal=<SignalRelay on Foo(x=8)>,
    args=(
        EmissionInfo(
            signal=<_DataclassFieldSignalInstance 'x' on <SignalGroup 'FooSignalGroup' with 1 signals>>,
            args=(8, 1),
            attr_name=None
        ),
    ),
    attr_name='foo'
)
```

tlambert03 commented:
@brisvag, you expressed interest in this a while back.  Not sure if it's still something you care about, but have a look at the API above if you're inclined.

@Czaki, if you want to review as well, it's welcome.

@ndxmrb as well, since you recently expressed interest, let me know if it would work for you.

brisvag commented:
Hey! Excited to see this picked up again :)

I did a quick-and-dirty test with old code from https://github.com/napari/napari/pull/4474. Maybe I converted things wrong, but I don't think it's working with containers a the moment.

I wasn't sure how to use the EventedModel (if it's possible) with nested eventes, so I switched to dataclasses like in your example.

```py
from dataclasses import dataclass, field
from psygnal import evented
from psygnal.containers import EventedList, EventedDict, EventedSet

@evented
@dataclass
class A:
    x: int = 1

@evented(connect_child_events=True)
@dataclass
class M:
    a: EventedList = field(default_factory=lambda: EventedList([1, 2, 3]))
    b: EventedDict = field(default_factory=lambda: EventedDict({'a': 1, 'b': 2}))
    c: EventedSet = field(default_factory=lambda: EventedSet({'x', 'y', 'z'}))
    d: A = field(default_factory=A)

m = M()
m.events.connect(print)

m.a[1] = 12
m.a = [0, 2]
m.b['a'] = 3
m.b = {'x': 11}
m.c.add('w')
m.c = {1}
m.d.x = 12
m.d = {'x': 1}
```

I would expect 8 events, but I only get 5:

```py
EmissionInfo(
    signal=<_DataclassFieldSignalInstance 'a' on <SignalGroup 'MSignalGroup' with 4 signals>>,
    args=([0, 2], EventedList([1, 12, 3])),
    attr_name=None
)
EmissionInfo(
    signal=<_DataclassFieldSignalInstance 'b' on <SignalGroup 'MSignalGroup' with 4 signals>>,
    args=({'x': 11}, EventedDict({'a': 3, 'b': 2})),
    attr_name=None
)
EmissionInfo(
    signal=<_DataclassFieldSignalInstance 'c' on <SignalGroup 'MSignalGroup' with 4 signals>>,
    args=({1}, EventedSet({'w', 'z', 'x', 'y'})),
    attr_name=None
)
EmissionInfo(
    signal=<SignalRelay on A(x=12)>,
    args=(EmissionInfo(signal=<_DataclassFieldSignalInstance 'x' on <SignalGroup 'ASignalGroup' with 1 signals>>, args=(12, 1), attr_name=None),),
    attr_name='d'
)
EmissionInfo(
	signal=<_DataclassFieldSignalInstance 'd' on <SignalGroup 'MSignalGroup' with 4 signals>>,
	args=({'x': 1}, A(x=12)),
	attr_name=None
)
```

tlambert03 commented:
K thanks.  This was mostly focusing on the models themselves... I'll go back and check all the evented containers too.  (We do eventually want them to work)

tlambert03 commented:
ok @brisvag, evented containers should be working now.  Added a test at `test_nested_containers.py`.  Note also that EmissionInfo objects now have a `flatten()` method, so your example, updated, would be:

```python
from dataclasses import dataclass, field

from rich import print

from psygnal import evented
from psygnal.containers import EventedDict, EventedList, EventedSet


@evented
@dataclass
class A:
    x: int = 1


@evented(connect_child_events=True)
@dataclass
class M:
    a: EventedList = field(default_factory=lambda: EventedList([1, 2, 3]))
    b: EventedDict = field(default_factory=lambda: EventedDict({"a": 1, "b": 2}))
    c: EventedSet = field(default_factory=lambda: EventedSet({"x", "y", "z"}))
    d: A = field(default_factory=A)


m = M()
m.events.connect(lambda e: print(e.flatten()))

m.a[1] = 12
m.a = [0, 2]
m.b["a"] = 3
m.b = {"x": 11}
m.c.add("w")
m.c = {1}
m.d.x = 12
m.d = {"x": 1}
```

prints

```python
EmissionInfo(
    signal=<SignalInstance 'changed' on <SignalGroup 'ListEvents' with 9 signals>>,
    args=(1, 2, 12),
    loc=('a', 'changed')
)
EmissionInfo(
    signal=<_DataclassFieldSignalInstance 'a' on <SignalGroup 'MSignalGroup' with 4 signals>>,
    args=([0, 2], EventedList([1, 12, 3])),
    loc=('a',)
)
EmissionInfo(
    signal=<SignalInstance 'changing' on <SignalGroup 'DictEvents' with 6 signals>>,
    args=('a',),
    loc=('b', 'changing')
)
EmissionInfo(
    signal=<SignalInstance 'changed' on <SignalGroup 'DictEvents' with 6 signals>>,
    args=('a', 1, 3),
    loc=('b', 'changed')
)
EmissionInfo(
    signal=<_DataclassFieldSignalInstance 'b' on <SignalGroup 'MSignalGroup' with 4 signals>>,
    args=({'x': 11}, EventedDict({'a': 3, 'b': 2})),
    loc=('b',)
)
EmissionInfo(
    signal=<SignalInstance 'items_changed' on <SignalGroup 'SetEvents' with 1 signals>>,
    args=(('w',), ()),
    loc=('c', 'items_changed')
)
EmissionInfo(
    signal=<_DataclassFieldSignalInstance 'c' on <SignalGroup 'MSignalGroup' with 4 signals>>,
    args=({1}, EventedSet({'z', 'w', 'x', 'y'})),
    loc=('c',)
)
EmissionInfo(
    signal=<_DataclassFieldSignalInstance 'x' on <SignalGroup 'ASignalGroup' with 1 signals>>,
    args=(12, 1),
    loc=('d', 'x')
)
EmissionInfo(
    signal=<_DataclassFieldSignalInstance 'd' on <SignalGroup 'MSignalGroup' with 4 signals>>,
    args=({'x': 1}, A(x=12)),
    loc=('d',)
)
```

tlambert03 commented:
> How does one enable this for EventedModel?

with patience young padawan :joy: 🧘

tlambert03 commented:
> Great! How does one enable this for `EventedModel`?

a more general and less sarcastic answer:
There's a bit of a divergence in the psygnal codebase at the moment.  all the dataclass stuff uses a single `SignalGroupDescriptor` object:

```python
@dataclass
class MyThing:
    events: ClassVar[SignalGroupDescriptor] = SignalGroupDescriptor()
```

and that one object is able to inspect the class on which it is defined and create a SignalGroup with events for all the fields.

The psygnal.EventedModel still uses the older napari-style metaclass structure, and therefore has some degree of logic duplication.  So, in the short time, implementing this for EventedModels means just going through and doing the corresponding stuff in the EventedModel that I've done here for SignalGroupDescriptor.  Ultimately, it would be even better to get EventedModel just using SignalGroupDescriptor itself.  That's possible, but we need to maintain support (at least at first) for all the property stuff on evented model.  There too, I think it would be good to check in with what pydantic has done in v2 to support properties, and ideally deprecate the features in psygnal's evented model.  

anyway, that's the general requirement

ndxmrb commented:
I was hoping for `EventedModel` to get this functionality, too. 😋 But it's nice to see this topic driven this fast.

Maybe I didn't fully understand the details of your explanation above, but I guess this is about computed fields? 

So providing something like the `@computed_field` decorator (https://docs.pydantic.dev/latest/api/fields/#pydantic.fields.computed_field), but addinitionally specifying the `field_dependencies`? Maybe it can be saved into `ComputedFieldInfo` (https://docs.pydantic.dev/latest/api/fields/#pydantic.fields.ComputedFieldInfo) or a subclass thereof...

tlambert03 commented:
mostly a note to self: this is more or less ready to go, but i want to reconsider what we're putting in the `loc` field, particularly for child events of nested containers, before merging

jni commented:
@tlambert03 just stumbled on this. It's exciting! Let me know if you want to pair up at some point in case you want someone to talk things out with.

tlambert03 commented:
@jni, the primary thing that I need to decide here is the API of that `loc` parameter.  it's a little bit overloaded at the moment, with a different meaning for evented *fields* (as on dataclasses or models) vs signals on an object that don't necessarily represent an addressable value:

for example, here is the same example as [above](https://github.com/pyapp-kit/psygnal/pull/298#issuecomment-1991708001), but with comments putting "what happened" into natural language.  Note the slight difference depending on the type of signal (`_DataclassFieldSignalInstance` signals change the value itself, which container objects change an item within the container)


```python
# item 1 of attribute a changed from 2 to 12
m.a[1] = 12
EmissionInfo(
    signal=<SignalInstance 'changed' on <SignalGroup 'ListEvents' with 9 signals>>,
    args=(1, 2, 12),
    loc=('a', 'changed')
)

# attribute a changed from [1, 12, 3] to [0, 2]
m.a = [0, 2]
EmissionInfo(
    signal=<_DataclassFieldSignalInstance 'a' on <SignalGroup 'MSignalGroup' with 4 signals>>,
    args=([0, 2], EventedList([1, 12, 3])),
    loc=('a',)
)

# item 'a' of attribute b changed from 1 to 3
m.b["a"] = 3
EmissionInfo(
    signal=<SignalInstance 'changing' on <SignalGroup 'DictEvents' with 6 signals>>,
    args=('a',),
    loc=('b', 'changing')
)
EmissionInfo(
    signal=<SignalInstance 'changed' on <SignalGroup 'DictEvents' with 6 signals>>,
    args=('a', 1, 3),
    loc=('b', 'changed')
)

# attribute b changed from {'a': 1, 'b': 2} to {'x': 11}
m.b = {"x": 11}
EmissionInfo(
    signal=<_DataclassFieldSignalInstance 'b' on <SignalGroup 'MSignalGroup' with 4 signals>>,
    args=({'x': 11}, EventedDict({'a': 3, 'b': 2})),
    loc=('b',)
)

# attribute c had item 'w' added to the set
m.c.add("w")
EmissionInfo(
    signal=<SignalInstance 'items_changed' on <SignalGroup 'SetEvents' with 1 signals>>,
    args=(('w',), ()),
    loc=('c', 'items_changed')
)

# attribute c was changed from {'z', 'w', 'x', 'y'} to {1}
m.c = {1}
EmissionInfo(
    signal=<_DataclassFieldSignalInstance 'c' on <SignalGroup 'MSignalGroup' with 4 signals>>,
    args=({1}, EventedSet({'z', 'w', 'x', 'y'})),
    loc=('c',)
)

# attribute x on attribute d changed from 1 to 12
m.d.x = 12
EmissionInfo(
    signal=<_DataclassFieldSignalInstance 'x' on <SignalGroup 'ASignalGroup' with 1 signals>>,
    args=(12, 1),
    loc=('d', 'x')
)

# attribute d changed from  {'x': 12} to {'x': 1}
m.d = {"x": 1}
EmissionInfo(
    signal=<_DataclassFieldSignalInstance 'd' on <SignalGroup 'MSignalGroup' with 4 signals>>,
    args=({'x': 1}, A(x=12)),
    loc=('d',)
)

```


I want to unify those things somehow so that the consumer doesn't need to do too much conditional logic every time to determine what actually happened.  It might require changing the signals on the containers (which would be breaking) or just changing the semantics of `loc` (which is totally new and added in this PR, and free to change)

your thoughts on that would be very welcome

jni commented on src/psygnal/_evented_decorator.py:
```diff
@@ -83,6 +86,14 @@ def evented(
+        also "evented" (as determined by the `is_evented` function in this module,
+        which returns True if the class has been decorated with `@evented`, or if it
+        has a SignalGroupDescriptor) to the group on the parent object. By default
+        False.
```
... True?

jni replied:
(Which btw I think should be the default... I think it takes a lot of experience with events to even notice that there's no bubbling up by default, let alone grok it. I think most users would expect bubbling to automagically work.)

jni commented on src/psygnal/_group.py:
```diff
@@ -45,11 +45,30 @@ class EmissionInfo(NamedTuple):
 
     signal: SignalInstance
     args: tuple[Any, ...]
+    loc: str | int | None | tuple[int | str, ...] = None
```
Looking at the earlier example, which had `attr_name` here, I find that easier to think about than `loc`. This is especially true if you add dictionaries into the mix — so now we don't know from the type whether this is an index, a key, or an attribute name. What do you think about having three optionals:
- attr
- index
- key

Of course, now I move on to read about "flatten"... 😅 

You could make a lightweight "Loc" object that could contain each of the keys, then you could have a sequence of Loc objects? The nice thing about this option is that you could write an accessor that gets you the right object given the parent and the sequence of locs.

I'm also not super keen on the "loc" name. "from"? "emitted_by"? "child"?

ndxmrb replied:
What about "origin"?

I think "loc" would be fine though...

tlambert03 commented:
Not annoying at all, and thanks for the ping @ndxmrb 

I often think of this pr as well and lament the fact that it didn't get in.  I agree it's nice functionality and not that far from being ready.


I got a little stymied by the loc stuff, and dealing with what I would have to break to give it the api i wanted.  But I shouldn't let that paralyze me 😅.  Let me try to pick it up again this weekend

tlambert03 commented:
## ✅ PR Ready for Merge - Final Status Update

This PR has been thoroughly tested and documented. Here's the complete status:

### 🎯 **Core Feature Status: COMPLETE**
- ✅ Event bubbling from nested evented children works perfectly
- ✅ Path tracking via `PathStep` and `EmissionInfo` implemented
- ✅ Support for dataclasses, evented containers (List, Dict, Set) 
- ✅ Clean API with `connect_child_events=True` parameter
- ✅ Backward compatibility maintained (no breaking changes)

### 🧪 **Test Coverage: EXCELLENT (99%)**
- ✅ All existing tests pass (331 passed)
- ✅ Comprehensive test coverage for new features
- ✅ Added `tests/test_path_coverage.py` with edge case testing
- ✅ Diff coverage: **99%** (only 1 untestable Python <3.10 line remaining)
- ✅ Coverage includes error conditions, edge cases, and all code paths

### 📚 **Documentation: COMPLETE**
- ✅ Added comprehensive event bubbling guide to `/docs/guides/dataclasses.md`
- ✅ Includes practical examples with nested dataclasses and containers
- ✅ Covers event paths, common use cases, and performance tips
- ✅ Documentation builds successfully without errors

### 🔍 **Code Quality**
- ✅ No breaking changes to existing APIs
- ✅ Clean, well-structured implementation
- ✅ All previous concerns addressed (path API design, default values, test coverage)
- ✅ Performance impact minimal and documented

### 📋 **Recent Additions (this review)**
- Added comprehensive test coverage for uncovered edge cases
- Created user-friendly documentation with practical examples  
- Validated feature robustness under various conditions
- Confirmed readiness for production use

**Recommendation: This PR is ready to merge.** The feature is well-implemented, thoroughly tested, and properly documented. It adds valuable functionality while maintaining backward compatibility.

Czaki commented:
reviewing

tlambert03 commented on src/psygnal/containers/_evented_dict.py:
```diff
@@ -110,22 +103,39 @@ def __get_pydantic_core_schema__(
     """Events available on [EventedDict][psygnal.containers.EventedDict]."""
 
-    adding = Signal(object)  # (key, )
+    adding = Signal(object, signal_instance_class=DictSignalInstance)  # (key, )
```
i did start with that, but that (I think) requires me to subclass `Signal`, which is not possible without relaxing mypyc on that class (and adding `@mypyc_attr(allow_interpreted_subclasses=True)`) ... which comes with a slight performance hit.  I can, however, do this:

```python
DictSignal = partial(Signal, signal_instance_class=DictSignalInstance)

class DictEvents(SignalGroup):
    """Events available on [EventedDict][psygnal.containers.EventedDict]."""

    adding = DictSignal(object)  # (key, )
```

will do that, but let me know if that wasn't your point

tlambert03 replied:
If it ends up being a commonly needed thing, we can consider adding an additional API for it.  But it's not necessary for this pr (and in fact, wouldn't work for evented list because they don't all want the same signal instance type)... so some other time

tlambert03 commented:
> also it looks like I cannot connect:
> 
> ```python
> team.events.leader.connect(lambda info: print(f"Event2: {info}"))
> ```
> 
> and get update of any event of leader field. But I need to connect to `team.events` and parse much more events.

`team.events.leader` is an event emitted when the value of the `leader` field on `team` is changed.  *not* when any value of `team.leader` is changed (`that would be `team.leader.events`).  So, if you stuck with the rest of the example, and used `team.leader.name = "Alice"` ... then I would not expect `team.events.leader` to have been emitted... here is a more thorough test (that I have now added) that clarifies all the things you should expect:

```python

from psygnal import testing, evented, EmissionInfo
from dataclasses import dataclass, field

def test_team_example():
    @evented
    @dataclass
    class Person:
        name: str = ""
        age: int = 0

    @evented
    @dataclass
    class Team:
        name: str = ""
        leader: Person = field(default_factory=Person)

    team = Team()

    # This will trigger the listener above
    team_level_info = EmissionInfo(
        team.leader.events.name,
        ("Alice", ""),
        path=(PathStep(attr="leader"), PathStep(attr="name")),
    )
    team_leader_level_info = EmissionInfo(
        team.leader.events.name, ("Alice", ""), path=(PathStep(attr="name"),)
    )
    with (
        testing.assert_emitted_once_with(team.events, team_level_info),
        testing.assert_emitted_once_with(team.leader.events, team_leader_level_info),
        testing.assert_emitted_once_with(team.leader.events.name, "Alice", ""),
        testing.assert_not_emitted(team.events.leader),
    ):
        team.leader.name = "Alice"

```

tlambert03 commented:
thanks for your review @Czaki, and I look forward to any follow up thoughts.

@ndxmrb, @brisvag,  and @jni.  it would be great to get your eyes on this one more time as well.  The easiest high-level entry point (if you need some context again) is probably the new section added to the documentation, [here](https://github.com/pyapp-kit/psygnal/blob/495b978c553f050cf7080ea166c6579cc2a8e046/docs/guides/dataclasses.md#event-bubbling-from-child-objects)

Czaki commented:
> I dunno, I think it's pretty useful :) if, with this comment, you were referring specifically to the `team.events.leader` confusion, then please see comment above.

I see how it works. 

But see example. If I want to be notified to any update of leader field form your example, I need to connect to all events of from object. `team.events` and construct callback like:

```python
def leader_update(event)
	if event.path[0].attr != '.leader':
		return
	# handle update of leader 
```

Or write own logic to handle
```python 
team.leader = new_leader
```

To disconnect the signal from the old leader and connect it to a new one?

Still, the first option with early return is easier to write, and many users may select it over performance.



> I did try very hard to keep performance the same. There were indeed some performance regressions at points here, but I fixed them in [1f1f99b](https://github.com/pyapp-kit/psygnal/commit/1f1f99be99bf1325dd5549d3bd9b23db201dfbee). You can see the resulting benchmarks in [codspeed.io/pyapp-kit/psygnal/branches/tlambert03%3Anested2](https://codspeed.io/pyapp-kit/psygnal/branches/tlambert03%3Anested2)

I see that you moved connection to sub events to be executed after the first callback is requested to connect. If every callback is disconnected, it may be nice to disconnect internal callbacks but I think that it may be a less common situation, than creating and updating objects without callbacks.

tlambert03 commented:
> But see example. If I want to be notified to any update of leader field form your example, I need to connect to all events of from object. team.events

No, you need to connect to team.leader.events.  Same as always

ndxmrb commented:
> I dunno, I think it's pretty useful :)

I think so, too! State patterns should be fairly easy with this, I think. Speaking of state pattern: I'm trying to create something like that using `EventedModel`, to see how it works and if docs are giving a good start.

At first, I had no problems with distinguishing what you said. But connecting to the correct signals when mixing evented containers and variables sometimes gets confusing after a while. Maybe you could add a section to the docs, so that users don't need to look into the test  or find the remark in this PR to get back on track again. 

Something like what you said above (formatting by me):

> `team.events.leader` is an event emitted when the value of the `leader` field on `team` is changed. _Not_ when any value of `team.leader` is changed (that would be `team.leader.events`).

tlambert03 commented:
> At first, I had no problems with distinguishing what you said. But connecting to the correct signals when mixing evented containers and variables sometimes gets confusing after a while.

Yeah, I do agree with that... but i think this might just be the nature of the nested problem.  but here's how I see it, and tell me whether this helps (or if you have suggestions for improved documentation):
- dataclasses are objects that have fields.  the `Foo()` class might have a field `bar: Bar`, and we access it with regular attribute access:  `foo = Foo(); foo.bar`
- evented dataclasses gain a new attribute `events` that is a SignalGroup containing multiple signals – one for each field on the dataclass – that emit when the corresponding field changes.  `foo.events.bar` is emitted when the value of `foo.bar` changes.
- if `Bar` is *itself* an evented dataclass, then *it too*, just like `Foo`, will have an `events` SignalGroup that emits whenever a field on `bar` changes: `bar.events`.  When accessed via the top level foo, this will look like `foo.bar.events`.
- These two are not to be confused:
    - `foo.events.bar` is the *`Signal`* that is emitted when the `bar` field on `foo` is changed (i.e. `foo.bar = ...`).
    - `foo.bar.events` is the *`SignalGroup`* that emits whenever any field on `bar` changes (i.e. `foo.bar.field = ...`).  In other words, as soon as you access `foo.bar`, the value you get back is a `Bar`... it has no concept of the fact that it lives on a `Foo` ... and therefore we shouldn't expect any of it's events to have any knowledge of a foo

tlambert03 commented on docs/guides/dataclasses.md:
```diff
@@ -323,6 +323,129 @@ class Person:
+    name: str = ""
+    age: int = 0
+
+@dataclass
+class Team:
+    events: ClassVar[SignalGroupDescriptor] = SignalGroupDescriptor(
+        connect_child_events=True
+    )
```
Torn here, I know that it's a little confusing to have two ways to do this, but for those who like having their linter not yell at them for using the name "events", I like showing this way of doing it.  I should probably add a comment to that effect (even though it's mentioned elsewhere on this page)

tlambert03 replied:
updated significantly.  includes mention of the default value now (True), and shows the two patterns in different tabs

tlambert03 commented:
Yeah, the two general points that I think are coming up frequently:

- if you connect to `foo.bar.events`, then it's important to realize that `foo` knows _nothing_ about that (because you accessed `foo.bar`.  And there's no way for foo to switch your callback over to a new bar if it changes.  That's just the nature of attributes not knowing their parents (this is to @ndxmrb last point)
- yes, `foo.events.bar` is specifically the event of setting the field bar on foo (as it had always been).  It is not "all events on `bar`, is "the event of changing `bar` on `foo`.  The semantics haven't changed here, but adding bubbling of events (from `foo.bar.events` up to `foo.events`) does now demand additional awareness by the user in the case of nested data classes.

Hopefully the docs I added make that clear.

tlambert03 commented:
> PS: can't resist noting: @evented is so much nicer than the manual thing. 😂

Nicer perhaps in how it looks, but worse for static typing and auto completion.  For reference, we need this in the typing world https://github.com/python/typing/issues/213

Without that sort of feature `@evented` will always just be a "cool" syntactic trick that comes at the cost of your IDE not knowing anything about the new "events" field.   The only reason IDEs know anything about `@dataclass` modifying a class is that it is special cased within type checkers

tlambert03 commented:
> My preferred solution would be, that my code snippet above does JustWork® by automatically dis-/reconnecting.

```python
model.list_of_models.events.connect(my_callback)
old_state = deepcopy(model)
model.update(old_state)
model.list_of_models.append(...)
```

this particular pattern is not likely to ever work.  consider that deepcopy makes a new instance of everything, that example can be recast as this:

```python
model.list_of_models.events.connect(my_callback)
model.list_of_models = EventedList()  # what update would effectively do
model.list_of_models.append(...)
```

and now, the key bit here is to recognize that you have used `model.list_of_models` in every case.  So let's recast it again:

```python
listA = model.list_of_models 
listA.events.connect(my_callback)
listA = EventedList()  # we have simply overwritten listA here
listA.append(...)
```

so ... the only way to make this work in the future would *not* be a method on `model.list_of_models`, like `model.list_of_models.events.connect(my_callback)` ... it would have to be a method on `model` itself.  So, if (and only if) you're specifically  in the situation where you don't want your callback to have to filter the field name, then we could add something like:

```python
model.events.connect(my_callback, restrict_to_specific_field='list_of_models')
```

tlambert03 commented:
> code readability is king, especially in Python

to an extent yes ... but I don't think there's a giant readability difference between

```python
@evented
@dataclass
class Thing:
    ...

# and 

@dataclass
class Thing:
    events: ClassVar = SignalGroupDescriptor()
```

but there is a major maintainability difference (with the second one being way easier to maintain).  So, while i think "readability is king" is an excellent first/primary principle.  I think it's import to not be so stuck on it that we don't weight costs benefits.  People sometimes get so excited by decorators (as neat syntactic sugar) that they prefer them just on coolness factor (i.e. `@evented` "ooooo" ... `events: ClassVar = SignalGroupDescriptor()` "meh...").

> but why wouldn't something like this work:

you're still thinking dynamically.  The point here is that a decorator can't tell a type checker that it has "added" a field.

```python
def add_field(cls: T) -> T:
    setattr(cls, 'new_field', object())

class Thing: ...

Thing = add_field(Thing)  # no way to say we've added `new_field` here

thing = Thing()
thing.new_field  # TypeError
```

jni commented:
after minor reflection:

```
events: ClassVar = AutomaticChangeNotifications()
```

or

```
events: ClassVar = CallbacksOnChange()
```
...?

not quite right but definitely less jargony. Something in that direction would help. Still don't know what to do about the fact that I don't know about ClassVars. I guess some targeted docs would help.

tlambert03 commented:
> I'm sorry I am having a hard time linking my code snippet (connecting one set of events to another) to yours (dynamically adding attributes)... What's the middle step I'm missing?

it's possible we've been talking about different things this whole time?  If we were discussing why "`@evented` is so much nicer than the manual thing"... then I wasn't quite sure how your code snippet related to that discussion:

> I haven't actually fished around in the implementation but why wouldn't something like this work:
> ```
> for field in self.fields:
>    getattr(self, field).events.connect(getattr(self.events, field))
> ```

it was mentioned in the context of https://github.com/python/typing/issues/213 ... but I don't really see how it connects, cause it's a runtime mechanism of connecting something, rather than a typing strategy.  It doesn't solve the fact that type checkers won't know that `@evented` adds an attribute named `.events`

> I'm hardly a Python noob and I actually don't know what ClassVar means. So I doubt I'm the only person who would be confused by this code.

Well, these are tricky scenarios... because that is the [official way that the documentation suggests](https://docs.python.org/3/library/dataclasses.html#class-variables) to indicate that a class attribute is not a field on the dataclass (i.e. *instances* don't have their own attribute field).  So, I don't really know what common denominator of having-read-the-docs I should be targeting?

> I realise that evented hides the same thing, but that's kind of the point: it shows exactly what the user needs to know and no more.

It don't think it does actually.  It indicates semantics *to a user who has already read the HTML documentation somewhere else and already knows what `@evented` means*.  Wheres `events: ClassVar = Thing()` tells me: "ok, this is adding a non-dataclass attribute, and I can access it by the name events`  That gives you both semantics *and* tells both the user and the IDE what attribute they can access that new functionality.

> There's no convention to my knowledge about where to put the events ClassVar in the attributes list,

I think this goes back to you needed to understand the significance of [ClassVars in standard library dataclasses](https://docs.python.org/3/library/dataclasses.html#class-variables).  You can put them anywhere, and call them anything.

> Also, I can easily insert a typo in `event` and the explicit form will work just fine: now the name events is just a convention rather than being programmatically determined.

this is a feature not a bug.  
A) lets you decide more explicitly what you want the `SignalGroup` to be called.
B) if you did use a typo (`my_obj.event`) then at runtime, *both* of these strategies would fail.  However, only the `ClassVar` version would be able to catch that bug statically.

jni commented:
There were two discussions in parallel:

1) evented decorator vs weirdass ClassVar. 😜 (genuinely kidding, thanks for the docs links)

2) the next step after this PR that many of us are pining for: how to automagically bubble up events in such a way that if I connect to obj.events.field, I am notified when obj.field.subfield changes. I *believe* this could be achieved at runtime by my event-to-event connection shenanigans. Am I wrong?

tlambert03 commented:
> that many of us are pining for: how to automagically bubble up events in such a way that if I connect to obj.events.field, I am notified when obj.field.subfield changes

I just want to make sure something is clear here.

The functionality you're after *is* already possible... just not using the syntax you proposed here (`obj.events.field.connect`).  As discussed in https://github.com/pyapp-kit/psygnal/pull/298#issuecomment-3005235858, with this PR, you can already achieve this exact functionality with one additional line:

```python
@obj.events.connect  # (not obj.events.field.connect)
def _on_field_changed(info: EmissionInfo):
    if info.signal.name == 'field':
        # go nuts
```

that will work, and will bubble up changes from `obj.field.subfield`, and will continue to work even if obj.field is re-assigned to a new instance or subfield is reassigned to a new instance.

Was that already clear?  and are you just pining for that small wrapper to be put into a public API?

jni commented:
> Was that already clear? and are you just pining for that small wrapper to be put into a public API?

Yes, that's clear (and again thank you for your many explanations above until we all grokked it 😂🙏), that's why I wrote:

> Since it's user-space-fixable, and it requires significant engineering effort, I do agree with merging this PR as-is and then making an issue for the magical connection-following.

My point is a bit more subtle:

1. (minor) the filtering could generate a large number of Python function calls compared to how many we actually want to listen to. It's unlikely that this will ever be a limiting factor but it feels a bit wasteful. 🤷 
2. because `obj.events.field` continues to exist (and obviously must continue to exist), it's an [attractive nuisance](https://blog.ganssle.io/articles/2023/01/attractive-nuisances.html), in that even experienced users of psygnal such as @Czaki, @ndxmrb and myself are tempted to subscribe to that instead of doing the filtering. I imagine many users will have a cursory look at the bubbling docs (or will not read the docs, rather learn about it from some comment somewhere), and will try to subscribe to `obj.events.field`, and will be surprised when they don't get notified when a subfield changes. (Think of the many times that we have got issues in napari from folks subscribing to `viewer.layers.events.changed` wanting to be notified when a layer is added or removed.)
  So, given all this, magical connection and disconnection *might* be a good thing to do for users.

Is *that* point clear? Again, I'm very happy to have the conclusion be "too hard/finicky" or "will mess with my static analysis so won't fix" or "here's an argument about event purity that you haven't considered", but I think it's important to consider *if* it's possible to do this well, then it might be a good feature.

tlambert03 commented:
Ahh I see.  You want to actually overload that connection, to keep its original semantics, but gain additional semantics of sub fields?

I appreciate the concept of an attractive nuisance here, but adding that would be more than just a convenient magic... it would be a big breaking change.  And that is because `obj.events.field` is (and always has been) an individual SignalInstance... not a SignalGroup.  As such, callbacks that connect to it must have a signature of (new value, old value)... _not_ a signature of (EmissionInfo).

So we can't just start calling those callbacks with sub fields packed into an EmissionInfo object.

Thoughts?

jni commented:
> As such, callbacks that connect to it must have a signature of (new value, old value)

Can we be even more magical so that when a (potentially deeply nested) sub-field changes, we still generate an old-value/new-value pair for the outer field, and keep the signature the same? 🤔 Or are you saying that we want the signature to change so folks can quickly tell what changed?

jni commented:
> In other words:
> obj.field.subfield = 42
> 
> Would call:
> 
> on_field_value_changed(obj.field)

Bingo!

tlambert03 commented:
Sorry that took me so long 😅

Yes, we can do something like that.  
I think, though, that I might prefer to have it gated by an argument to connect(), like connect(..., emit_on_evented_mutations=True), or something like that.

This kind of thing could also work well for, eg, the evented proxy pattern used with numpy arrays: to simply say to the callback "you should have another look at this object, it has somehow changed"

And, if we really need to know exactly what has changed in this scenario (lacking Emission info in the callback), the Signal.current_emitter() pattern could perhaps be extended to include that info

ndxmrb commented:
I agree, happy to see it merged, great discussion! 

This code snipped works nicely for field changes.
```
@obj.events.connect  # (not obj.events.field.connect)
def _on_field_changed(info: EmissionInfo):
    if info.signal.name == 'field':
        # go nuts
```
But I think (if the change from https://github.com/pyapp-kit/psygnal/pull/298#issuecomment-3016672697 comes) it wouldn't be that easy for nested structures. That's why I proposed something to get the `info.path` string-comparable more easily or even use [glob pattern](https://en.wikipedia.org/wiki/Glob_(programming)). 

```
@obj.events.connect  # (not obj.events.field.connect)
def _on_field_changed(info: EmissionInfo):
    if str(info.path) == 'field.subfield': # string compare
        # go nuts
    elif info.path.islike('field[0].subfield': # glob pattern if field would be a list
        # go ham on first element
    elif info.path.islike('field[*].subfield': # glob pattern again
        # go full ham
```

tlambert03 commented:
> Why? Can't you put that in the path just like normal?

Because, and this is the really important thing to remember here... `obj.events.field` is _not_. `SignalGroup`, it is a `SignalInstance`.  And therefore the callback does not receive an EmissionInfo object (and _thats_ the thing that has a `path`).  It receives an instance of `obj.field`

brisvag commented:
Ah, right, I kept forgetting that also while reading the discussion. Tricky indeed!

brisvag commented:
Yep, I get that. And I expect that as we use this in the core codebase of projects like napari, we'll be fine. I just know that downstream users will intuitively (as did I many a time) reach for `obj.field.events.connect` instead of `obj.events.field.connect`, and then will get surprised when the callbacks stop working after they do `obj.field = new`.

I understand that at some point the only solution is "read the docs and do it right", but it's also true that the first option *feels* right at a first glance.

jni commented:
I've made #375 just to keep track of this feature request. As discussed above, there are user-side solutions while that gets implemented, so no rush fix.
